Model/src/main/java/org/gusdb/wdk/model/fix/table/edaanalysis/plugins/VDIMigrationPlugin.java

```diff
@@ -1,89 +1,153 @@
+  }
+
+  // Visible for testing
+  void readVdiMappingFile(File mappingFile) {
+    _legacyIdToVdiId = readLegacyStudyIdToVdiId(mappingFile);
+  }
+
   @Override
   public TableRowInterfaces.RowResult<AnalysisRow> processRecord(AnalysisRow nextRow) throws Exception {
     final String legacyDatasetId = nextRow.getDatasetId();
+
     if (!legacyDatasetId.startsWith(UD_DATASET_ID_PREFIX)) {
-      return new TableRowInterfaces.RowResult<>(nextRow)
-          .setShouldWrite(false);
+      return new TableRowInterfaces.RowResult<>(nextRow).setShouldWrite(false);
     }
 
     final String legacyUdId = legacyDatasetId.replace(UD_DATASET_ID_PREFIX, "");
-    final String vdiId = legacyIdToVdiId.get(legacyUdId);
+    final String vdiId = _legacyIdToVdiId.get(legacyUdId);
 
     if (vdiId == null) {
       LOG.warn("Unable to find legacy ID " + legacyUdId + " in the tinydb file.");
       missingFromVdiCount.incrementAndGet();
-      return new TableRowInterfaces.RowResult<>(nextRow);
+      return new TableRowInterfaces.RowResult<>(nextRow)
+          .setShouldWrite(false);
     }
 
     // Append UD prefix to VDI ID. The prefix is prepended in the view that maps stable VDI IDs to the unstable study
     // ID, which is the currency of EDA.
     final String vdiDatasetId = UD_DATASET_ID_PREFIX + vdiId;
-    nextRow.setDatasetId(vdiDatasetId);
+    final Optional<String> vdiEntityId = _vdiEntityIdRetriever.queryEntityId(vdiDatasetId);
+    if (vdiEntityId.isEmpty()) {
+      LOG.warn("Unable to find entity ID in appdb for VDI dataset ID: " + vdiDatasetId);
+      return new TableRowInterfaces.RowResult<>(nextRow)
+          .setShouldWrite(false);
+    }
+
+    LOG.info("Analysis descriptor before migration: " + nextRow.getDescriptor());
+    String descriptor = nextRow.getDescriptor().toString();
+
+    // Find all variable IDs.
+    final Set<String> legacyVariableIds = VAR_ID_PATTERN.matcher(descriptor).results()
+        .map(match -> match.group(1))
+        .collect(Collectors.toSet());
+
+    final String entityId = ENTITY_ID_PATTERN.matcher(descriptor).results()
+        .findAny()
+        .map(m -> m.group(1))
+        .orElse(null);
+
+    // Replace all entityID with entityID looked up from database.
+    if (entityId != null) {
+      descriptor = descriptor.replaceAll(entityId, vdiEntityId.get());
+    }
+
+    // Replace all variable IDs with value converted from legacy variable ID.
+    for (String legacyVariableId: legacyVariableIds) {
+      descriptor = descriptor.replaceAll(legacyVariableId, convertToVdiId(legacyVariableId));
+    }
 
-    return new TableRowInterfaces.RowResult<>(nextRow)
+    // Create a copy with just the dataset ID updated to VDI counterpart.
+    AnalysisRow out = new AnalysisRow(nextRow.getAnalysisId(), vdiDatasetId, new JSONObject(descriptor),
+        nextRow.getNumFilters(), nextRow.getNumComputations(), nextRow.getNumVisualizations());
+    
+    LOG.info("Analysis descriptor after migration: " + out);
```

Comment:
You're dumping the AnalysisRow here; might have a decent toString() but isn't necessarily good for comparing to the "before" log line above.